src/mpid/ch4/shm/xpmem/xpmem_avl.h

```diff
@@ -0,0 +1,564 @@
+        sure all of sub tree still obey the AVL tree requirement (left
+        branch and right branch height difference cannot exceed 1). If
+        it exceeds this limit, we need to adjust the height of tree to
+        make it balance and satisfy the requirement of AVL tree again.
+
+    MPIDI_XPMEM_seg_avl_tree_delete_all:
+        This function checks reference count of node and output warning
+        info if reference count is larger than 0, then it deletes and
+        releases all the node buffer pools allocated.
+
+*/
+
+
+/****************************************/
+/* AVL tree init related functions */
+/****************************************/
+#undef FUNCNAME
+#define FUNCNAME MPIDI_XPMEM_seg_avl_tree_init
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_seg_avl_tree_init(MPIDI_XPMEM_avl_tree_t * tree)
+{
+    int i;
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_SEG_AVL_TREE_INIT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_SEG_AVL_TREE_INIT);
+
+    tree->root = NULL;
+    tree->tree_size = 0;
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_XPMEM_SEG_AVL_TREE_INIT);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+
+/****************************************/
+/* AVL tree delete related functions */
+/****************************************/
+#undef FUNCNAME
+#define FUNCNAME MPIDI_XPMEM_avl_tree_node_ref_check
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_avl_tree_node_ref_check(MPIDI_XPMEM_avl_tree_t * tree)
+{
+    int ret;
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_XPMEM_seg_t *node_stack[64];
+    int addr_stack[64];
+    int sp = 0;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_AVL_TREE_NODE_REF_CHECK);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_AVL_TREE_NODE_REF_CHECK);
+
+    if (tree->tree_size) {
+        MPIDI_XPMEM_seg_t *node = tree->root;
+        while (1) {
+            if (node->left != NULL) {
+                node_stack[sp] = node;
+                addr_stack[sp++] = 0;   /* 0 stands for recovering from LEFT. */
+                node = node->left;
+                continue;
+            }
+          left_recovery:
+            if (node->right != NULL) {
+                node_stack[sp] = node;
+                addr_stack[sp++] = 1;   /* 1 stands for recovering from RIGHT. */
+                node = node->right;
+                continue;
+            }
+          right_recovery:
+            ret = xpmem_detach((void *) node->vaddr);
+            MPIR_ERR_CHKANDJUMP(ret == -1, mpi_errno, MPI_ERR_OTHER, "**xpmem_detach");
+            if (node->ref_count > 1) {
+                printf("[Warning] XPMEM segment [%lu, %lu] hasn't been released\n", (uint64_t)node->vaddr,
+                       (uint64_t) node->vaddr + node->size);
+                fflush(stdout);
```

Comment:
You should not directly use `printf`. I'd just call `MPIR_Assert(node->ref_count == 0);`. It fails only in an incorrect user program, and it is an `no-op` if user chooses fast build (i.e., `--enable-fast`).